In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("/Users/noahschutz/datascience/userssharedsdfschoolimprovement2010grants.csv")

### Having imported our data, we need to take a look at it to see what we can extract from it.

In [3]:
print(data)

                                          School Name               City  \
0                   HOGARTH KINGEEKUK MEMORIAL SCHOOL           SAVOONGA   
1                                     AKIACHAK SCHOOL           AKIACHAK   
2                                      GAMBELL SCHOOL            GAMBELL   
3                                BURCHELL HIGH SCHOOL            WASILLA   
4                                        AKIAK SCHOOL              AKIAK   
5                                      MIDVALLEY HIGH            WASILLA   
6                                     TULUKSAK SCHOOL           TULUKSAK   
7                         CAPITOL HEIGHTS JR HIGH SCH         MONTGOMERY   
8                               LOWNDES CO MIDDLE SCH       FORT DEPOSIT   
9                                  CENTRAL MIDDLE SCH           ROCKFORD   
10                            BELLINGRATH JR HIGH SCH         MONTGOMERY   
11                              HAYNEVILLE MIDDLE SCH         HAYNEVILLE   
12          

### So our dataset has 7 columns: school name, city, state, district, grant amount, model selected, and location.

### Our employer is comparing these schools based on the amount they are granted, which is our most interesting variable, as its the one that appears to be calcuated from values in the other columns. If we are to draw a conclusion that our employer can use, we need to assume that the grant amount is a dependent variable influneced by the independent variables in other columns.

### I'm going to look at a few basic calculations just to get a feel for my data.

In [4]:
data.shape

(831, 7)

### Confirmed 7 columns = confirmed 7 variables. Also we have 831 total observations for schools.

In [5]:
data.describe()

School Name          City State         District Name  \
count                          831           831   831                   831   
unique                         821           424    50                   421   
top     MOUNT PLEASANT HIGH SCHOOL  PHILADELPHIA    CA  PHILADELPHIA CITY SD   
freq                             2            29    92                    27   

       2010/11/Award Amount  Model Selected  \
count                   757             825   
unique                  651               4   
top                   $2.00  Transformation   
freq                     17             608   

                                                 Location  
count                                                 831  
unique                                                823  
top     6941 NORTH CENTRAL ST\nPORTLAND, OR 97203\n(45...  
freq                                                    3

### It appears we might have a few duplicates in our midst, but to make sure, we should review our data. Pandas allows us to search for duplicates with a single command.

In [6]:
data.duplicated()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
801    False
802    False
803    False
804    False
805    False
806    False
807    False
808    False
809    False
810    False
811    False
812    False
813    False
814    False
815    False
816    False
817    False
818    False
819    False
820    False
821    False
822    False
823    False
824    False
825    False
826    False
827    False
828    False
829    False
830    False
Length: 831, dtype: bool

In [7]:
data.drop_duplicates()

School Name               City  \
0                   HOGARTH KINGEEKUK MEMORIAL SCHOOL           SAVOONGA   
1                                     AKIACHAK SCHOOL           AKIACHAK   
2                                      GAMBELL SCHOOL            GAMBELL   
3                                BURCHELL HIGH SCHOOL            WASILLA   
4                                        AKIAK SCHOOL              AKIAK   
5                                      MIDVALLEY HIGH            WASILLA   
6                                     TULUKSAK SCHOOL           TULUKSAK   
7                         CAPITOL HEIGHTS JR HIGH SCH         MONTGOMERY   
8                               LOWNDES CO MIDDLE SCH       FORT DEPOSIT   
9                                  CENTRAL MIDDLE SCH           ROCKFORD   
10                            BELLINGRATH JR HIGH SCH         MONTGOMERY   
11                              HAYNEVILLE MIDDLE SCH         HAYNEVILLE   
12                               SOUTHLAWN MIDDLE SCH         MONTGOMERY   
13                            CONTINUOUS LEARNING CTR             MOBILE   
14                            JACKSON-STEELE ELEM SCH         HAYNEVILLE   
15                                MCINTYRE MIDDLE SCH         MONTGOMERY   
16                                JOHN ESSEX HIGH SCH          DEMOPOLIS   
17                                   CALHOUN HIGH SCH        LETOHATCHEE   
18                              OSCEOLA MIDDLE SCHOOL            OSCEOLA   
19                                   HALL HIGH SCHOOL        LITTLE ROCK   
20                            ROSE CITY MIDDLE SCHOOL  NORTH LITTLE ROCK   
21                           TRUSTY ELEMENTARY SCHOOL         FORT SMITH   
22                              J.A. FAIR HIGH SCHOOL        LITTLE ROCK   
23                                OSCEOLA HIGH SCHOOL            OSCEOLA   
24                           CLOVERDALE MIDDLE SCHOOL        LITTLE ROCK   
25                          GADSDEN ELEMENTARY SCHOOL            GADSDEN   
26                            EAST VALLEY HIGH SCHOOL               MESA   
27                           CRITTENTON YOUTH ACADEMY            PHOENIX   
28                               ALCHESAY HIGH SCHOOL         WHITERIVER   
29                             SAN CARLOS HIGH SCHOOL         SAN CARLOS   
..                                                ...                ...   
801                          WASHINGTON MIDDLE SCHOOL             YAKIMA   
802  FOSTER & WILLIAMS HIGH SCHOOL OF THE VISUAL ARTS          MILWAUKEE   
803                                       DUBOIS HIGH          MILWAUKEE   
804                                       CUSTER HIGH          MILWAUKEE   
805                     JAMES MADISON ACADEMIC CAMPUS          MILWAUKEE   
806                                      VINCENT HIGH          MILWAUKEE   
807                                     BAY VIEW HIGH          MILWAUKEE   
808       DOWNTOWN INSTITUTE OF ARTS & LETTERS (DIAL)          MILWAUKEE   
809                               SOUTH DIVISION HIGH          MILWAUKEE   
810         MILWAUKEE AFRICAN AMERICAN IMMERSION HIGH          MILWAUKEE   
811                WHS LAW EDUCATION & PUBLIC SERVICE          MILWAUKEE   
812                           BRADLEY TECHNOLOGY HIGH          MILWAUKEE   
813                                       HAMLIN PK-8             HAMLIN   
814                             RIVERSIDE HIGH SCHOOL              BELLE   
815                DODDRIDGE COUNTY ELEMENTARY SCHOOL         WEST UNION   
816                            MOUNT VIEW HIGH SCHOOL              WELCH   
817                            WEST HAMLIN ELEMENTARY        WEST HAMLIN   
818                          MALDEN ELEMENTARY SCHOOL         CHARLESTON   
819                   MARTINSBURG NORTH MIDDLE SCHOOL        MARTINSBURG   
820                         SANDY RIVER MIDDLE SCHOOL           AVONDALE   
821                   STONEWALL JACKSON MIDDLE SCHOOL         CHARLESTON   
822           

### Since no entries were dropped, I'm going to assume that some schools just so happen to be named the same. Either that, or someone enterred their school name twice, but accidentally typoed something the second time because they are so skillful with accidentally leaving their hand on the keyboard without realizing it.

### Our data also has a few entries that don't record a grant amount. If these missing values were recorded in a column that was independent and unnecesary to our conlcusions then we would probably have to fill them in ourselves. However, since we're looking for patterns in the "grant amount" column itself, those empty rows aren't really that useful to us. We can just get rid of any of the columns with missing values, and we'll still have plenty of entries that we can use to find trends in our data.

In [8]:
data = data[pd.notnull(data['2010/11/Award Amount'])]

In [9]:
data.describe()

School Name          City State         District Name  \
count                757           757   757                   757   
unique               748           382    46                   372   
top     EAST HIGH SCHOOL  PHILADELPHIA    CA  PHILADELPHIA CITY SD   
freq                   2            29    92                    27   

       2010/11/Award Amount  Model Selected  \
count                   757             757   
unique                  651               4   
top                   $2.00  Transformation   
freq                     17             543   

                                                 Location  
count                                                 757  
unique                                                749  
top     6941 NORTH CENTRAL ST\nPORTLAND, OR 97203\n(45...  
freq                                                    3

### We still have 757 entries that we can analyze using graphics resources such as plot.ly. 

In [10]:
data.loc[:,"2010/11/Award Amount"].mean

<bound method Series.mean of 0      $471014.00
1      $520579.00
2      $449592.00
3      $641184.00
4      $399686.00
5      $697703.00
6      $427658.00
7      $110544.00
8       $12973.00
9      $199828.00
10     $165903.00
11     $121657.00
12     $914557.00
13     $198953.00
14     $118394.00
15     $143238.00
16     $151116.00
17     $163619.00
18     $665000.00
19     $198728.00
20     $991451.00
21     $990135.00
22     $197012.00
23     $695000.00
24     $198783.00
25     $847003.00
26     $381512.00
27     $915959.00
28     $812195.00
29     $928654.00
          ...    
801    $115268.00
802      $5000.00
803      $5000.00
804    $650093.00
805    $749320.00
806    $928024.00
807    $134909.00
808      $5000.00
809    $124643.00
810    $567905.00
811      $5000.00
812      $1202.00
813    $258600.00
814    $481900.00
815    $464200.00
816    $450000.00
817    $156660.00
818    $217598.00
819    $461690.00
820    $370000.00
821    $241211.00
822    $314618.00
823    $389906.00

### This gives us information, but it doesn't allow us to find any trends in our data. Plot.ly can help with this.

In [11]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly
import plotly.plotly as py
import plotly.graph_objs as go
plotly.tools.set_credentials_file(username='ns248', api_key='6XA3DRSaSVsboozfFxZF')
init_notebook_mode(connected=True)

### What potential commands could be useful for us to plot? In other words, what would a potential client desire from our manipulating of the data?

### Let's start by histogramming all of it.

In [12]:
dataHist = [go.Histogram(y=data['2010/11/Award Amount'])]

In [13]:
iplot(dataHist)

### Cool, so our data has a ton of entries where the grant size is either close to 0 dollars or bettwen 100,000 and 200,000 dollars. Maybe we can find out whether schools with certain locations recieve consistently variable amounts of grant money.

In [14]:
data.dtypes # I do this because we're going to need to convert one of our varaibles to a different type

School Name             object
City                    object
State                   object
District Name           object
2010/11/Award Amount    object
Model Selected          object
Location                object
dtype: object

In [15]:
data['2010/11/Award Amount'] = data['2010/11/Award Amount'].str.replace('$', '')

In [16]:
data[['2010/11/Award Amount']] = data[['2010/11/Award Amount']].astype(float) #this converts the monetary string into a number we can graph

In [17]:
data.groupby('State')['2010/11/Award Amount'].median()

State
AK    471014.0
AL    151116.0
AR    665000.0
AZ    728696.0
CA    169199.5
CO    619609.0
CT    590000.0
DC    481370.0
DE    651608.0
FL    652965.0
GA    142540.0
IA        18.0
ID    248340.0
IL    175105.0
IN    199941.0
KS    144190.5
KY    437283.0
MA    665108.0
MD    773922.0
MI    198255.0
MN    561111.0
MO    379991.5
MS     70576.0
MT    259182.0
NC    693605.0
ND    418574.0
NE    510953.0
NH    346179.0
NJ    163269.0
NM        15.0
NV    500345.0
NY    117372.0
OH    451750.0
OK    134920.0
OR    616667.0
PA    483332.5
SC    579145.0
SD    265445.0
TN    290710.5
UT    549444.0
VA    499793.0
VT    250173.5
WA    170768.5
WI    124643.0
WV    341250.0
WY    297300.0
Name: 2010/11/Award Amount, dtype: float64

### Well those semantics took way too long to figure out. Had to replace the dollar signs in order to convert the grant amounts to floats. Anyway, now we can accurately find numerical statistics for the amount column.

In [18]:
dataBar = [go.Bar(x=data['State'], y=data['2010/11/Award Amount'])]

In [19]:
iplot(dataBar)

### which gets us total amount of grant money per state. But since this doesn't take into account how many schools recieve money per state, and states with more schools represented will probably recieve more money, we need to take the average instead and incorporate that into our chart.

In [20]:
dataAver = data.groupby(['State']).mean()

### I used groupby to get the average grant amount of a school per state. Now let's plot this.

In [21]:
dataAverSort=dataAver.sort_values('2010/11/Award Amount')
dataAverSort

2010/11/Award Amount
State                      
MS            156269.125000
NJ            179858.583333
IA            190308.750000
AL            209162.000000
KS            211776.666667
NM            219463.333333
IL            219711.000000
CA            220508.956522
NY            226754.040000
ID            248340.000000
SD            265445.000000
VT            275048.300000
OK            277048.500000
WI            288736.000000
IN            313109.571429
MT            326948.333333
WV            330739.533333
WY            334227.333333
GA            342035.730769
WA            364956.166667
NH            380229.142857
MI            389058.214286
MO            412973.343750
TN            416475.666667
ND            418574.000000
DC            419933.900000
OH            436193.885714
VA            444943.888889
OR            482669.000000
PA            488244.948276
NV            499156.800000
KY            502593.800000
AK            515345.142857
CO            534643.789474
MD            548507.000000
AR            562301.285714
CT            574366.928571
NE            582582.714286
NC            588136.434783
SC            593300.368421
UT            607375.000000
FL            607589.563380
MN            612284.052632
MA            620105.500000
DE            651608.000000
AZ            693755.000000

In [22]:
dataAverSortN=dataAverSort #creating an alternate array so I can graph with State on the x-axis

In [23]:
dataAverSortN.index.name = 'State'
dataAverSortN.reset_index(inplace=True) 

In [24]:
iplot(
    go.Figure(
        data = [
            go.Bar(x=dataAverSortN['State'], y=dataAverSort['2010/11/Award Amount'])],
    
        layout = go.Layout(
            height = 500,
            title = 'School Grant Size per State',
            xaxis=dict(
            title='State'),
            yaxis=dict(
            title='Average Amount of Grant ($)')
        )
))

### And now we have graphed the average grant size per state! However we're also concerned about organizing by model type. Let's make a plot for this too. (At least this will have fewer columns)

In [25]:
dataInterAver = data.groupby(['Model Selected']).mean() #The inter stands for intervention, as in intervention model

In [26]:
dataInterAverSort=dataInterAver.sort_values('2010/11/Award Amount')
dataInterAverSort

2010/11/Award Amount
Model Selected                      
Closure                 75155.500000
Turnaround             394100.581818
Transformation         434116.062615
Restart                503412.424242

In [27]:
dataInterAverSortN=dataInterAverSort #creating an alternate array again

In [28]:
dataInterAverSortN.index.name = 'Intervention Model'
dataInterAverSortN.reset_index(inplace=True) 
dataInterAverSortN

Intervention Model  2010/11/Award Amount
0            Closure          75155.500000
1         Turnaround         394100.581818
2     Transformation         434116.062615
3            Restart         503412.424242

In [29]:
iplot(
    go.Figure(
        data = [
            go.Bar(x=dataInterAverSortN['Intervention Model'], y=dataInterAverSortN['2010/11/Award Amount'])],
    
        layout = go.Layout(
            height = 500,
            title = 'Average Grant Size per Intervention Model',
            xaxis=dict(
            title='Type of Intervention Model Used'),
            yaxis=dict(
            title='Avereage Amount of Grant ($)')
        )
))

### If we had more columns than this I'd probably try a few more graph calculations, but until I get more information I'm comfortable talking about the information I currently have. 